packages/ui/src/components/va-form/VaForm.vue

```diff
@@ -8,115 +9,120 @@
+      onChildUnmounted: (removableChild: FormChild) => childUnmountedHandler(removableChild),
+    })
 
-@Options({
-  name: 'VaForm',
-  emits: ['validation'],
-})
-export default class VaForm extends mixins(
-  FormPropsMixin,
-) {
-  nestedFormElements: (FormChild | VaForm)[] = [];
-
-  parentFormProvider = setup(() => {
-    return { ...inject(FormServiceKey, undefined) }
-  })
-
-  formProvider = setup(() => {
-    const onChildMounted = (child: FormChild | VaForm) => this.childMountedHandler(child)
-    const onChildUnmounted = (removableChild: FormChild | VaForm) => this.childUnmountedHandler(removableChild)
-
-    const formProvider = {
-      onChildMounted,
-      onChildUnmounted,
+    const childMountedHandler = (child: FormChild) => {
+      nestedFormElements.value.push(child)
     }
 
-    provide(FormServiceKey, formProvider)
+    const childUnmountedHandler = (removableChild: FormChild) => {
+      nestedFormElements.value = nestedFormElements.value.filter(child => child !== removableChild)
+    }
 
-    return formProvider
-  })
+    /** @public */
+    const reset = () => {
+      nestedFormElements.value
+        .filter(({ reset }) => reset)
+        .forEach((item) => { item.reset() })
+    }
 
-  childMountedHandler (child: FormChild | VaForm) {
-    this.nestedFormElements.push(child)
-  }
+    const resetValidation = () => {
+      nestedFormElements.value
+        .filter(({ resetValidation }) => resetValidation)
+        .forEach((item: any) => { item.resetValidation() })
+    }
 
-  childUnmountedHandler (removableChild: FormChild | VaForm) {
-    this.nestedFormElements = this.nestedFormElements.filter(child => child !== removableChild)
-  }
+    const focus = () => {
+      const focusableElement = nestedFormElements.value.find(({ focus }) => focus)
+      if (focusableElement) {
+        focusableElement.focus()
+      }
+    }
 
-  mounted () {
-    this.parentFormProvider.onChildMounted?.(this)
+    const focusInvalid = () => {
+      const invalidComponent = nestedFormElements.value
+        .find((item) => !isVaForm(item) && item.hasError && item.hasError())
 
-    if (this.autofocus) {
-      this.$nextTick(() => { this.focus() })
-    }
-  }
-
-  unmounted () {
-    this.parentFormProvider.onChildUnmounted?.(this)
-  }
-
-  /** @public */
-  reset () {
-    this.nestedFormElements
-      .filter(({ reset }) => reset)
-      .forEach((item) => { item.reset() })
-  }
-
-  resetValidation () {
-    this.nestedFormElements
-      .filter(({ resetValidation }) => resetValidation)
-      .forEach((item: any) => { item.resetValidation() })
-  }
-
-  focus () {
-    const focusableElement = this.nestedFormElements.find(({ focus }) => focus)
-    if (focusableElement) {
-      focusableElement.focus()
+      if (invalidComponent) {
+        invalidComponent.focus()
+      } else {
+        nestedFormElements.value
+          .forEach(item => isVaForm(item) && item.focusInvalid?.())
+      }
```

Comment:
If we have
`const isVaForm = (value: any) => !!value.focusInvalid`
So we don't need to check for the existence of the method:
`item.focusInvalid?.()` ==> `item.focusInvalid()`